Hash-Table

d-tiapkin opened the pull request.

d-tiapkin committed:
First iteration

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+#include <list>
+#include <vector>
+
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> >
```
Начиная с C++11 можно писать `>>`.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+
+    // CONSTRUCTORS
+
+    explicit HashMap() {
```
Зачем explicit пустому конструктору? Это нарушит такой кейс:

```
HashMap<...> func() {
    if (...) return {};
    ...
}
```

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        return it;
+    }
+
+    const size_t _START_SIZE = 4;
```
Идентификаторы, начинающиеся с _ и заглавной буквы, по стандарту использовать нельзя: https://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier

На мой взгляд, более хорошая привычка -- ставить подчёркивания в конце членов класса. А в константах они вообще не нужны (`constexpr size_t INITIAL_SIZE = 4;`)

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    explicit HashMap() {
+        _table.resize(_START_SIZE);
+    }
+
```
Лишняя пустая строка

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    using CNode = std::pair<const KeyType, ValueType>;
+    using Node =  std::pair<KeyType, ValueType>;
+
+    // CONSTRUCTORS
```
```
                                                                                    
                               m                           m                        
  mmm    mmm   m mm    mmm   mm#mm   m mm  m   m   mmm   mm#mm   mmm    m mm   mmm  
 #"  "  #" "#  #"  #  #   "    #     #"  " #   #  #"  "    #    #" "#   #"  " #   " 
 #      #   #  #   #   """m    #     #     #   #  #        #    #   #   #      """m 
 "#mm"  "#m#"  #   #  "mmm"    "mm   #     "mm"#  "#mm"    "mm  "#m#"   #     "mmm" 
                                                                                    
                                                                                    
```

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    HashMap(std::initializer_list<Node> lst) {
+        _table.resize(2*lst.size());
```
Откуда магическая константа 2?

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+
+    // CONSTRUCTORS WITH HASHERS
+
+    explicit HashMap(Hash hs) :
```
Hash hash

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    // INSERT-ERASE
+
```
Зачем столько пустых строк?

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        return _size == 0;
+    }
+
+    auto hash_function() const {
```
Если уж ты сделал названия рубрик, почему это в рубрике size?

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    }
+
+private:
+    struct TableValue {
```
По смыслу скорее TableEntry.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        _table.resize(new_size);
+    }
+
+    void normalise() {
```
normalize

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    void normalise() {
+        if (_fullness > _table.size() * 3/4) {
```
Магические константы нужно назвать и вынести туда же, куда и START_SIZE. Конкретно эта штука называется load factor.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+            size_t n = _table.size();
+            std::list<CNode> copy(_lst);
+
+            rebuild(2*n);
```
А это -- grow factor. (или growth, погугли, что чаще встречается).

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    void normalise() {
+        if (_fullness > _table.size() * 3/4) {
+            size_t n = _table.size();
+            std::list<CNode> copy(_lst);
```
Тут можно std::move использовать, а не копировать лишний раз, я не прав?

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+
+            rebuild(2*n);
+
+            for (auto it = copy.begin(); it != copy.end(); ++it) {
```
for (const auto& it:  copy)

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        typename std::list<CNode>::iterator it;
+    };
+
+    void rebuild(size_t new_size) {
```
По названию непонятно, почему эта функция очищает хештаблицу.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    size_t find_key(const KeyType& x) const {
+        size_t it = _hasher(x)%(_table.size());
```
Пара пробелов вокруг % просятся.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        size_t it = _hasher(x)%(_table.size());
+
+        while (_table[it].deleted || (_table[it].used && !(_table[it].it->first == x))) {
+            it = (it + 1)%(_table.size());
```
Возможно, быстрее будет так:

```
if (++it == _table.size()) it = 0;
```

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        bool was_deleted = _table[it].deleted;
+
+        _lst.push_back(x);
+        _table[it] = {true, false, --_lst.end()};
```
Стоит сделать так, чтобы было понятно, что это за параметры:

```
_table[it] = { /* used = */ true, /* deleted = */ false, --lst.end() };
```

А ещё лучше сделать TableEntry конструктор от итератора.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    void add(const Node& x, bool normal) {
```
Зачем параметр normal? Во-первых, стоит назвать его shouldNormalize; во-вторых, не правда ли, что если он false, то ты гарантируешь, что размер подходящий и что normalize ничего не сделает?

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+
+private:
+    struct TableValue {
+        bool used;
```
Я бы сделал `enum class State { FREE, USED, DELETED };`.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    const size_t _START_SIZE = 4;
+
+    std::vector<TableValue> _table;
+    std::list<CNode> _lst;
```
Не надо сокращать list до lst.

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+        HashMap copy(oth);
+
+        rebuild(copy._table.size());
+        for (auto it : copy) {
```
const auto&

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    template<typename Iter>
+    HashMap(Iter begin, Iter end, Hash hs) :
```
Можно в каждый конструктор добавить параметром хешер по умолчанию, тем самым сократив число конструкторов в два раза:

```template<typename Iter>
HashMap(Iter begin, Iter, end, Hash hash = Hash{})
```

ifsmirnov commented on contest2/hash_table.h:
```diff
@@ -0,0 +1,259 @@
+    HashMap(std::initializer_list<Node> lst) {
+        _table.resize(2*lst.size());
+        for (auto it : lst) {
+            add(it, true);
```
1) почему здесь add, а в предыдущем методе insert?
2) почему в этом конструкторе нельзя вызвать generic от пары итераторов?

ifsmirnov reviewed (changes requested):
Ещё немного про копипасту в конструкторах.

d-tiapkin committed:
Some changes

 #####      #######    #            #   #######          #####       #           #           #        #           #     #####       #######    #####  
#         #    #           #   # #      # #   #                     #         #     #           #        #   #      # #        #   #         #      #               #         # 
#               #           #   #   #  #   #   #                     #                #           #      #       #    #   #      #   #                 #               #       
 #####      #           #   #     #     #   #####             #                #######     #          #   #     #    #   #     ####    #####        #####  
           #    #           #   #            #   #                     #                #           #    #######   #       #  #   #          #     #                           # 
#        #     #           #   #            #   #                     #          #    #           #    #          #   #        # #   #          #     #       #                  # 
 #####      #######    #            #   #######          #####       #           #    #          #   #           #     #####      #######      #####
